javalib/src/main/scala/java/util/IdentityHashMap.scala

```diff
@@ -0,0 +1,136 @@
+ */
+
+package java.util
+
+import java.{util => ju}
+
+import IdentityHashMap.IdentityBox
+
+/**
+ * Does not support Double.NaN keys.
+ */
+class IdentityHashMap[K, V](inner: HashMap[IdentityBox[K], V])
+    extends AbstractMap[K, V] with Map[K, V] with Serializable with Cloneable {
+
+  def this(expectedMaxSize: Int) =
+    this(new HashMap[IdentityBox[K], V](expectedMaxSize, HashMap.DEFAULT_LOAD_FACTOR))
+
+  def this() =
+    this(HashMap.DEFAULT_INITIAL_CAPACITY)
+
+  def this(initialMap: java.util.Map[_ <: K, _ <: V]) = {
+    this(initialMap.size())
+    putAll(initialMap)
+  }
+
+  override def clear(): Unit = inner.clear()
+
+  override def clone(): AnyRef =
+    new IdentityHashMap(inner.clone().asInstanceOf[HashMap[IdentityBox[K], V]])
+
+  override def containsKey(key: Any): Boolean =
+    inner.containsKey(IdentityBox(key))
+
+  override def containsValue(value: Any): Boolean = inner.containsValue(value)
+
+  override def get(key: Any): V = inner.get(IdentityBox(key))
+
+  override def isEmpty(): Boolean = inner.isEmpty
+
+  override def put(key: K, value: V): V = inner.put(IdentityBox(key), value)
+
+  override def remove(key: Any): V = inner.remove(IdentityBox(key))
+
+  override def size(): Int = inner.size
+
+  override def values(): Collection[V] = inner.values()
+
+  override def entrySet(): Set[Map.Entry[K, V]] = new EntrySet
+
+  // views
+  private final class EntrySet extends AbstractSet[Map.Entry[K, V]] {
+
+    def iterator(): Iterator[Map.Entry[K, V]] = {
+      new ju.Iterator[Map.Entry[K, V]] {
+        val iter = inner.entrySetIterator()
+
+        def hasNext(): Boolean = iter.hasNext()
+
+        def next(): Map.Entry[K, V] = {
+          val e = iter.next()
+          new MapEntry(e)
+        }
+
+        def remove(): Unit = iter.remove()
+      }
+    }
+
+    def size(): Int = inner.size()
+
+    override def contains(value: Any): Boolean =
+      value match {
+        case value: Map.Entry[_, _] =>
+          val mapVal = inner.get(IdentityBox(value.getKey()))
+          mapVal == value.getValue()
+        case _ => false
+      }
+
+    override def remove(value: Any): Boolean =
+      value match {
+        case value: Map.Entry[_, _] =>
+          val mapVal = inner.remove(IdentityBox(value.getKey()))
+          mapVal == value.getValue()
+        case _ => false
+      }
+
+    override def clear(): Unit =
+      inner.clear()
+  }
+
+  private final class MapEntry(entry: Map.Entry[IdentityBox[K], V]) extends Map.Entry[K, V] {
+
+    override def equals(other: Any): Boolean =
+      other match {
+        case other: Map.Entry[_, _] =>
+          (entry.getKey() eq IdentityBox(other.getKey())) &&
+          Objects.equals(entry.getValue, other.getValue)
+        case _ => false
+      }
+
+    def getKey(): K = entry.getKey().inner
+
+    def getValue(): V = entry.getValue()
+
+    override def hashCode(): Int = Objects.hashCode(entry.getKey.inner) ^ Objects.hashCode(entry.getValue)
```

Comment:
This should use `System.identityHashCode(getKey())` instead of `Objects.hashCode`. Otherwise it isn't consistent with `equals`.

Reply:
I reused the `IdentityBox` code.